Refactor map layout

Jesus82 opened the pull request:
Refactors map layout, removes general 100% height chain of rules and fixes spacing

Jesus82 committed:
css fixes

lucaslcode commented on app/frontend/src/features/search/SearchResult.tsx:
```diff
@@ -65,14 +63,13 @@ export default function SearchResult({
               <div>
                 <Hidden smDown>
-                  <CouchIcon className={classes.statusIcon} />
+                  <CouchIcon className={classes.statusIcon}/>
```
Remember to run `yarn format`

lucaslcode commented on app/frontend/src/features/search/SearchPage.tsx:
```diff
@@ -15,14 +15,18 @@ const useStyles = makeStyles((theme) => ({
     [theme.breakpoints.up("md")]: {
-      flexDirection: "row",
+      display: "grid",
+      gridTemplateColumns: "30rem 1fr",
```
For some reason, the maplibre gl canvas doesn't fill the full space anymore until a user is selected or the window resized. Seems like it might be a maplibre gl bug? Might have to leave it as flex.

lucaslcode replied:
Also, why grid? You prefer setting the widths in one location?

Jesus82 replied:
I will take a look to that. It seems to me a JS initializing bug of the maplibre canvas.

Jesus82 replied:
Finally I used flex for the main layout, as it was the only way for the map to fit its space. Don't know why tbh (some JS black magic).

lucaslcode commented on app/frontend/src/features/search/SearchResult.tsx:
```diff
@@ -27,16 +27,14 @@ const useStyles = makeStyles((theme) => ({
   },
   statusLabelWrapper: {
-    display: "flex",
+    display: "grid",
+    gridTemplateColumns: "1fr 1fr",
```
I think flex is better for 1-dimensional layouts (with `justifyContent: "space-around"` to fix the spacing as you have done), but it's not a big deal

Jesus82 replied:
The _one dimension-flex many dimensions-grid_ rule is kind of popular, but I would say that is not very accurate. I recommend the beginning of this (great) course where that is discussed https://cssgrid.io/. Here grid let's you set the spacing and gutters in the same place (avoiding setting different margins), and also it is useful in order to control word break.

lucaslcode commented on app/frontend/src/features/search/SearchResult.tsx:
```diff
@@ -65,14 +63,13 @@ export default function SearchResult({
       <CardActionArea>
         <CardContent>
-          <UserSummary user={user} avatarIsLink={false} compact>
+          <UserSummary user={user} avatarIsLink={false}>
```
`compact` makes the avatar smaller and displays only the name. I think we don't need age and location here because there is a map to show location, and the age is listed below. People have annoyingly long location names (which we should maybe fix by not including state and country by default in the location editing...)

Jesus82 replied:
I see. Problem here is that compact outputs the small-sized avatar, when here the size should be the default. Should I write a method for this? (maybe a compact-text prop)?

lucaslcode replied:
Maybe another `nameOnly` prop?

lucaslcode replied:
Or rather separate `compact` into `nameOnly` and `smallAvatar`

lucaslcode commented on app/frontend/src/features/search/SearchPage.tsx:
```diff
@@ -15,14 +15,18 @@ const useStyles = makeStyles((theme) => ({
     flexDirection: "column-reverse",
-    height: "100%",
+    position: "fixed",
+    top: "4rem",
```
this should be `theme.shape.navPaddingXs` and then `theme.shape.navPaddingSmUp` at the right breakpoint.

lucaslcode reviewed (changes requested):
Thanks for tidying up the spacing!

Jesus82 committed:
Merge branch 'develop' of https://github.com/Couchers-org/couchers into frontend/refactor/full-container-in-maps

Jesus82 committed:
changed avatar props and fixed map glitch

Jesus82 committed:
fixes for mobile

Jesus82 commented:
Last commit addresses some issues in the mobile view. I set the scrolling to vertical (as now it is possible to do so, and makes cards more readable in mobile taking full width), but I could revert it back to horizontal easily.

Jesus82 committed:
format

Jesus82 committed:
removed fix in general rules

lucaslcode commented on app/frontend/src/features/search/SearchPage.tsx:
```diff
@@ -16,13 +16,14 @@ const useStyles = makeStyles((theme) => ({
-      gridTemplateColumns: "30rem 1fr",
+    [theme.breakpoints.up("sm")]: {
+      top: theme.shape.navPaddingSmUp,
+      display: "flex",
```
```suggestion
```

lucaslcode commented on app/frontend/src/features/search/SearchResultsList.tsx:
```diff
@@ -22,19 +22,28 @@ const useStyles = makeStyles((theme) => ({
   scroller: {
     marginTop: theme.spacing(3),
-    "&&": { alignItems: "flex-start" },
+    alignItems: "flex-start",
```
Sometimes the && is needed to increase the specificity to override the Material-UI style. At the moment this line doesn't get applied (but it doesn't matter for vertical scrolling)

lucaslcode reviewed (changes requested):
It should be horizontal scrolling on mobile - the whole page should scroll if we're doing vertical scroll, otherwise it's a bit unusual. (Also that component is called `HorizontalScroller`, so if we did want vertical scrolling we would need some refactoring.)

Jesus82 committed:
fixes horizontal scrolling and snap scrolling added

Jesus82 committed:
format

Jesus82 committed:
fixes

Jesus82 commented:
Shame on me I didn't realize the name of the component :D I've revert it back to horizontal and fixed spacing issues and added scroll snapping.

Jesus82 committed:
recovered padder to fix scroll-snapping

Jesus82 committed:
fix for text overflow in mobile cards

Jesus82 committed:
fix snap padding

Jesus82 committed:
fix snap padding

Jesus82 committed:
fix for text overflow in mobile cards

Jesus82 committed:
better fix for overflow

Jesus82 committed:
format

lucaslcode commented on app/frontend/src/features/search/SearchResultsList.tsx:
```diff
@@ -44,11 +49,17 @@ const useStyles = makeStyles((theme) => ({
+    marginBottom: theme.spacing(3),
+    borderRadius: theme.shape.borderRadius * 2,
+    backgroundColor: theme.palette.primary.contrastText,
+    boxShadow: "0 0 4px rgba(0,0,0,0.25)",
```
I think we should either use `elevation` on the card, or set this in the theme somewhere

Jesus82 replied:
Box shadow in MUI works only accepting parameters. So we cannot just set that boxShadow in the theme file, as it gives an error. The alternative would be to store it as a CSS variable (we can place that in the index.css file).

lucaslcode replied:
You can put what you like in the theme file (even just this whole string) if you extend the type definition (as is being done at the top of `theme.ts` for `theme.shape.borderRadius`)

lucaslcode commented on app/frontend/src/features/search/SearchResultsList.tsx:
```diff
@@ -44,11 +49,17 @@ const useStyles = makeStyles((theme) => ({
+    borderRadius: theme.shape.borderRadius * 2,
+    backgroundColor: theme.palette.primary.contrastText,
+    boxShadow: "0 0 4px rgba(0,0,0,0.25)",
+    [theme.breakpoints.down("md")]: {
+      "& > div > *": {
+        padding: 0,
+        margin: theme.spacing(2),
+        height: "calc(100% - 2rem)",
+        overflow: "hidden",
+      },
```
I don't understand this - there doesn't seem to be any div directly under the search result.

Jesus82 replied:
Don't know what happened as I had it working, but you're right. Switched it to button and it works.

lucaslcode replied:
I see. Can you use the MUI class instead of this selector - it's clearer what you're targeting. So it would be `"& MuiCardContent-root:`

lucaslcode commented on app/frontend/src/features/search/SearchResultsList.tsx:
```diff
@@ -21,21 +21,26 @@ import hasAtLeastOnePage from "utils/hasAtLeastOnePage";
   searchDesktop: {
-    margin: theme.spacing(0, 2),
-    marginTop: theme.spacing(2),
+    "& > div > *": {
+      flexShrink: 0,
+      width: "100%",
+    },
```
The field should accept a `fullWidth` prop which is probably clearer than doing this

lucaslcode commented on app/frontend/src/features/search/SearchResultsList.tsx:
```diff
@@ -44,11 +49,17 @@ const useStyles = makeStyles((theme) => ({
-      margin: theme.spacing(2),
+    marginBottom: theme.spacing(3),
+    borderRadius: theme.shape.borderRadius * 2,
+    backgroundColor: theme.palette.primary.contrastText,
```
`theme.palette.background.paper`?

lucaslcode commented on app/frontend/src/components/HorizontalScroller.tsx:
```diff
@@ -8,23 +8,29 @@ import useOnVisibleEffect from "../utils/useOnVisibleEffect";
-      marginInlineStart: theme.spacing(2),
+      flexShrink: 0,
+      height: "calc(100% - 2rem)",
+      width: "85vw",
+      maxWidth: "33rem",
```
I thought it's better to let the children decide their width

Jesus82 replied:
Without a set width, the card would grow as large as the text inside, making it a really long div.

lucaslcode replied:
This component takes `children` and the children can set their width. For example, the events in the community page fit a few in one screen, even on mobile, but now there is only one fully visible

lucaslcode reviewed (changes requested):
This is great, and I love the scroll snapping!

However, the horizontal scroller is also used on the community page and it looks a bit broken now (links: [this pr](https://frontend-refactor-full-container-in-maps--frontend.preview.couchershq.org/community/1/world), [develop](https://develop--frontend.preview.couchershq.org/community/1/world))

Jesus82 committed:
fix to places

Jesus82 committed:
fullWidth as prop

Jesus82 committed:
fix theme

Jesus82 commented:
I fixed when in community it scrolled the text when no places. But I don't know if that works with actual places, as there are none created.

Jesus82 committed:
format

Jesus82 committed:
Merge branch 'develop' into frontend/refactor/full-container-in-maps

lucaslcode commented:
> I fixed when in community it scrolled the text when no places.

Sorry - I was referring to the events

lucaslcode commented on app/frontend/src/components/HorizontalScroller.tsx:
```diff
@@ -8,23 +8,35 @@ import useOnVisibleEffect from "../utils/useOnVisibleEffect";
   root: {
-    "& > *": {
-      flex: "0 0 auto",
+    "& > div": {
```
Why did you change from * to div? What if the children are not divs?

lucaslcode commented on app/frontend/src/features/search/SearchResultsList.tsx:
```diff
@@ -44,11 +45,20 @@ const useStyles = makeStyles((theme) => ({
+      "& > button > *": {
+        padding: 0,
+        margin: theme.spacing(2),
+        height: "calc(100% - 2rem)",
```
```suggestion
        height: `calc(100% - ${theme.spacing(1)})`,
```
However I wonder if it's easier for the parent to have a padding, and the height to just be 100%. Or even better, `alignItems: "stretch"`?

darrenvong commented on app/frontend/src/components/HorizontalScroller.tsx:
```diff
@@ -8,23 +8,29 @@ import useOnVisibleEffect from "../utils/useOnVisibleEffect";
     WebkitOverflowScrolling: "touch",
+    overflowX: "scroll",
+    scrollSnapType: "x mandatory",
+    scrollPadding: theme.spacing(1.5),
```
I think we probably want to avoid using `0.5` units in the spacing value, so we should go with 1 or 2? Otherwise we are just moving off the 8px scale we agreed with design (even though I know they haven't been the best at sticking to it in the Figma designs) and defeats the point of having one.

Jesus82 replied:
I totally agree to follow the grid, but in some cases, the `4px` unit will be the only way to do that (for example if you apply `margin-left` and `margin-right` to a set of elements in order to achieve an `8px` gutter while still having them centered. Here it is more a visual matter. If we apply `scroll-padding: 16px`, `the box-shadow` of the previous card enters in the screen (in the left part of the scroller), which is not intended.
![Captura de pantalla de 2021-06-06 13-01-43](https://user-images.githubusercontent.com/8875832/120922228-2e735000-c6c8-11eb-816a-15e956a19897.png)

But if we apply `scroll-padding: 12px`, we get rid of that piece of shadow, and the proportion is still nice (3:4, which is used all over the design).
![Captura de pantalla de 2021-06-06 13-01-21](https://user-images.githubusercontent.com/8875832/120922148-cf154000-c6c7-11eb-99d7-8879d459da7e.png)

darrenvong replied:
Hmm maybe we should review the scale to be `4px` based instead. I'm seeing more and more cases where we have to break out of it because things don't look right otherwise, which suggests to me it's not granular enough or designs need to adjust to having more spaces around things...

Or stick to what we have and allow half units, but no more than that? Like you said, for anything that needs centering with margin/padding, the half unit seems like a necessary evil.

lucaslcode replied:
I think allow half in special circumstances. Having 4px grid would mean we would have lots of `theme.spacing(4)` which seems too high for the most common spacing.

Jesus82 committed:
Merge branch 'develop' into frontend/refactor/full-container-in-maps

Jesus82 committed:
fixed horizontal scroller in Community page

Jesus82 commented:
Made it so the width is determined just for the map scroller. The only pending thing would be the box shadow (I couldn't make that work in the theme file).

Jesus82 committed:
breakpoint fix

lucaslcode commented on app/frontend/src/features/communities/CommunityPage/CommunityPage.tsx:
```diff
@@ -19,33 +19,16 @@ export const useCommunityPageStyles = makeStyles((theme) => ({
   },
   cardContainer: {
-    alignItems: "flex-start",
-    [theme.breakpoints.down("xs")]: {
-      //break out of page padding
-      left: "50%",
-      marginLeft: "-50vw",
-      marginRight: "-50vw",
-      position: "relative",
-      right: "50%",
-      width: "100vw",
-    },
```
Not a fan of this? It allowed the horizontal scroller to go all the way to the edge

Jesus82 replied:
ooops, missed that. Will do another commit later this night.

lucaslcode commented on app/frontend/src/features/search/SearchResultsList.tsx:
```diff
@@ -44,11 +45,27 @@ const useStyles = makeStyles((theme) => ({
+      "&:last-child": {
+        marginRight: 0,
+      },
+      "& MuiCardContent-root": {
+        padding: 0,
+        height: `calc(100% - ${theme.spacing(4)})`,
+        overflow: "hidden",
+        margin: theme.spacing(2),
+      },
```
Invalid selector, looks fine though so remove this?

Jesus82 replied:
It was a fix for overlap happening in narrow screens. Should be working with last commit.

Jesus82 committed:
fix margin

lucaslcode reviewed (changes requested):
All looks good! Just some tiny things, and can you also fix the storybook example for `HorizontalScroller` which is broken on mobile somehow?

lucaslcode commented:
With the box shadow, I saw there's a few other places with a custom shadow so let's leave it for now.

Jesus82 committed:
fixes CSS

lucaslcode commented on app/frontend/src/components/HorizontalScroller.tsx:
```diff
@@ -13,7 +13,6 @@ const useStyles = makeStyles((theme) => ({
     display: "inline-flex",
     flexDirection: "row",
     height: "100%",
-    width: "100%",
```
scroll snap stops working with this.
I think you have to leave this in and do `"& > *": { flexShrink: 0 }`

lucaslcode replied:
@Jesus82 just needs the flex shrink 0 to fix storybook

lucaslcode commented:
Sorry about the merge conflicts - I tried to keep changes as minimal as possible. I'm happy to do it if you want.

Jesus82 committed:
merge

Jesus82 committed:
update button margins and fixed scroll snapping

Jesus82 commented:
> Sorry about the merge conflicts - I tried to keep changes as minimal as possible. I'm happy to do it if you want.

No worries, I think right now everything should be OK (I also adapted the spacing to fit the button in all cases.

Jesus82 committed:
fix for storybook and fix for text ellipsis

lucaslcode commented on app/frontend/src/features/search/SearchResult.tsx:
```diff
@@ -39,9 +44,24 @@ const useStyles = makeStyles((theme) => ({
+        display: "grid",
+        gridTemplateColumns: "1.25rem 1fr",
+        gridGap: theme.spacing(1),
+      },
```
```suggestion
        alignItems: "center",
      },
```

lucaslcode commented on app/frontend/src/features/search/SearchResult.tsx:
```diff
@@ -79,34 +102,19 @@ export default function SearchResult({
-              </Hidden>
-              <Typography
-                className={classes.statusLabel}
-                display="inline"
-                variant="body1"
-                color="primary"
-              >
+              <CouchIcon className={classes.statusIcon} />
+              <Typography display="inline" variant="body1" color="primary">
                 {hostingStatusLabels[user.hostingStatus]}
               </Typography>
             </div>
             <div>
-              <Hidden smDown>
-                <LocationIcon className={classes.statusIcon} />
-              </Hidden>
-              <Typography
-                className={classes.statusLabel}
-                display="inline"
-                variant="body1"
-              >
+              <LocationIcon className={classes.statusIcon} />
+              <Typography display="inline" variant="body1">
```
Thanks for fixing the wrapping, but I think we should still hide the icons for xsDown. We want to support 320px width displays and with the icons at that width, you can't see much label. (That might mean this needs to be flex rather than grid.)

Jesus82 replied:
This one is really tricky.
If you get rid of the icons, the labels lose their context, and in some cases is hard to guess what are they referring to

![Image 1](https://i.ibb.co/gzWyGFC/Captura-de-pantalla-de-2021-06-08-08-24-18.png)

Making it flex is also problematic, as when both texts don't fit, they push all the elements down, breaking the layout

![Image 2](https://i.ibb.co/260HRJf/Captura-de-pantalla-de-2021-06-08-08-28-09.png)

This could be even worse with languages that are not as economic as English when it comes to hosting.

![Image 3](https://i.ibb.co/k42zF3X/Captura-de-pantalla-de-2021-06-08-08-29-06.png)

My suggestion would be to have semantic icons (someting like a cross over the place icon meaning I can't host), and leave the text with ellipsis as a complement.

lucaslcode replied:
Good idea. Let's leave this for now then and ask design for some icons

Jesus82 replied:
Fine, let me know if there's something pending here

lucaslcode reviewed (changes requested):
Pinky promise this is the last one!

Jesus82 committed:
Update app/frontend/src/features/search/SearchResult.tsx

Co-authored-by: Lucas Levin <lucas@couchers.org>
